### PR TITLE
Add subjects handler to get all subjects in the current source

### LINKED ISSUE #30: Get all subjects in a document
This may already be possible, but I haven't been able to get anything to work.

I'm trying to get all the subjects in a given document, or even all the subjects matching some predicate in a given document.

Example of what I am trying to do:
```js
for await (const activity of data['https://jshurmer.inrupt.net/public/activities']) {
  //activity is now an ldflex object of each activity at that path
}
//or
for await (const bookmark of data['https://jshurmer.inrupt.net/public/bookmarks']) {
 //bookmark is an ldflex of each bookmark
}
//etc.
```

I'd be happy to contribute, if this currently is not supported, but could use a pointer on the best way to introduce this functionality. The sparql handler doesn't seem to to be handling this correctly, it expects a predicate but only has the url as a subject. Could this be fixed with a small change to that handler to support no-predicates? or Would it be better to introduce a new `.all` property that builds a different sparql entirely perhaps?

### PR DESCRIPTION
**Pull request by JordanShurmer**
This should resolve #30.

I could use some help getting this change tested properly, I can't seem to get this (nor query-ldflex) to build locally, which is the only way I know of to test this in an actual application. Tried writing some tests for them, but I'm not sure how exactly this can be tested without an actual (or mock) queryEngine.

I figured I'd go ahead and let people look at this change though, because it doesn't feel like quite the right solution, but this is my first change in here so I'm not entirely sure.

Thanks!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -80,6 +80,21 @@ async function showPerson(person) {
 }
 ```
 
+
+### Get all subjects from a document
+```javascript
+const typeIndex = path.create({ subject: namedNode('https://jshurmer.inrupt.net/settings/publicTypeIndex.ttl') });
+showTypes(typeIndex);
```
Comment (RubenVerborgh) [inline README.md]: Maybe we can have a different example? The subjects/types link might be confusing.

File: src/SubjectsHandler.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Queries for all subjects in the current queryEngine source
+ */
+export const SELECT_ALL_SUBJECTS = 'SELECT distinct ?thing WHERE { ?thing ?pred ?type }';
```
Comment (RubenVerborgh) [inline src/SubjectsHandler.js]: Let's just inline (only used once).
Also, consistent caps in `DISTINCT`, and perhaps just `?s ?p ?o` or `?subject ?predicate ?object`.

File: src/SubjectsHandler.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    const { queryEngine } = pathData.settings;
+    if (!queryEngine)
+      throw new Error(`${pathData} has no queryEngine setting`);
+
+    // simply execute the query and yield the results
+    for await (const bindings of queryEngine.execute(SELECT_ALL_SUBJECTS))
+      yield this.extractTerm(bindings, pathData);
+  }
+
+  extractTerm(binding, pathData) {
```
Comment (RubenVerborgh) [inline src/SubjectsHandler.js]: Code duplication of `ExecuteQueryHandler`, probably want to inherit from it.
Or, alternatively, we can change the `.sparql` handler instead, such that the `DISTINCT` query is generated there.

File: src/SubjectsHandler.js (hunk 2 above)
Comment (RubenVerborgh) [reply src/SubjectsHandler.js]: And definitely not export, this is an internal matter.

File: test/integration/execute-query-test.js [hunk 4]
```diff
@@ -75,6 +76,22 @@ describe('a query path with a path expression handler', () => {
...
+  beforeAll(() => {
+    const pathProxy = new PathProxy({ handlers: { ...handlersPath, subjects: new SubjectsHandler() }, resolvers });
+    person = pathProxy.createPath({ queryEngine }, { subject });
+  });
+
+  it('returns results, with executeQuery being called with the right query', async () => {
+    const subjects = [];
+    for await (const subj of person.subjects)
+      subjects.push(subj);
+    expect(queryEngine.execute).toBeCalledWith(SELECT_ALL_SUBJECTS);
```
Comment (RubenVerborgh) [inline test/integration/execute-query-test.js]: Good, let's just not import `SELECT_ALL_SUBJECTS` but inline it here.

File: test/unit/SubjectsHandler-test.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import SubjectsHandler from '../../src/SubjectsHandler';
+
+function toString() {
```
Comment (RubenVerborgh) [inline test/unit/SubjectsHandler-test.js]: Let's inline.

File: test/unit/SubjectsHandler-test.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  return 'path';
+}
+
+describe('a SubjectsHandler instance', () => {
```
Comment (RubenVerborgh) [inline test/unit/SubjectsHandler-test.js]: Maybe also have test on the actual workings here? (cfr. integration test)

**Comment by RubenVerborgh** [comment]
Good progress, thanks!

> I could use some help getting this change tested properly

Your tests seem good to me!

> I can't seem to get this (nor query-ldflex) to build locally

What's going wrong? Feel free to open an issue.

> but I'm not sure how exactly this can be tested without an actual (or mock) queryEngine.

Just like you did, that's perfect.

> I figured I'd go ahead and let people look at this change though, because it doesn't feel like quite the right solution, but this is my first change in here so I'm not entirely sure.

Alternative is implementing it in `.sparql`, then you reuse the (already tested) query engine logic.

File: src/SubjectsHandler.js (hunk 3 above)
Comment (JordanShurmer) [reply src/SubjectsHandler.js]: updating `.sparql` was my initial approach, but it was becoming more complicated than I thought. I'll keep thinking about it though, that smells like a better solution in the long run

**Comment by JordanShurmer** [comment]
@RubenVerborgh thanks for the feedback!

I think I've addressed all the issues. Can you take another look?

**Comment by RubenVerborgh** [comment]
Thanks, will also work on #32 (which did take the SPARQL route) and see if this one should go through SPARQL as well. In any case, tests and functionality will stay.

**Comment by RubenVerborgh** [comment]
Just merged #32, could you have a look and see if this influences anything?

**Comment by RubenVerborgh** [comment]
Actually I'll have a look to see if I can get that done.

**Comment by JordanShurmer** [comment]
Thanks @RubenVerborgh !

**Pull request merged**